server/src/main/java/org/opensearch/identity/ExtensionTokenProcessor.java

```diff
@@ -14,26 +37,80 @@
     public static final String INVALID_TOKEN_MESSAGE = "Token must not be null and must be a colon-separated String";
     public static final String INVALID_EXTENSION_MESSAGE = "Token passed here is for a different extension";
+    public static final String INVALID_ALGO_MESSAGE = "Failed to create a token because an invalid hashing algorithm was used.";
+    public static final String INVALID_PRINCIPAL_MESSAGE = "Token passed here is for a different principal.";
+
+    public static final String ALGORITHM = "AES";
+    public static final int KEY_SIZE = 128;
+    public static final int INITIALIZATION_VECTOR_SIZE = 96;
+    public static final int TAG_LENGTH = 128;
+    
+    private final byte[] extensionUniqueId;
+
+    private final byte[] initializationVector; 
+    private final SecretKey secretKey;  
 
-    private final String extensionUniqueId;
 
     public ExtensionTokenProcessor(String extensionUniqueId) {
-        this.extensionUniqueId = extensionUniqueId;
+        this.extensionUniqueId = extensionUniqueId.getBytes();
+        this.initializationVector = generateRandomIV(INITIALIZATION_VECTOR_SIZE);;
+        try {
+            this.secretKey = generateSecretKey(KEY_SIZE);
+        }
+        catch (NoSuchAlgorithmException noAlgo) {
+            
+            noAlgo.printStackTrace();
+            throw new Error(noAlgo);
+        }
     }
 
-    public String getExtensionUniqueId() {
-        return extensionUniqueId;
+    public static byte[] generateRandomIV(int ivLength){
+
+        byte[] nonce = new byte[ivLength/8];
+        SecureRandom secureRandom = new SecureRandom();
+        secureRandom.nextBytes(nonce);
+        return nonce;
     }
 
+    public static SecretKey generateSecretKey(int keyLength) throws NoSuchAlgorithmException {
+
+        KeyGenerator keyGen = KeyGenerator.getInstance(ALGORITHM);
+        keyGen.init(keyLength);
+        return keyGen.generateKey();
+    }
+    
+
     /**
      * Create a two-way encrypted access token for given principal for this extension
+     * @param: principal being sent to the extension
      * @return token generated from principal
+     * @throws NoSuchPaddingException
+     * @throws NoSuchAlgorithmException
+     * @throws InvalidAlgorithmParameterException
+     * @throws InvalidKeyException
+     * @throws BadPaddingException
+     * @throws IllegalBlockSizeException
+     * @throws IOException
      */
-    public PrincipalIdentifierToken generateToken(Principal principal) {
-        // This is a placeholder implementation
-        // More concrete implementation will be covered in https://github.com/opensearch-project/OpenSearch/issues/4485
-        String token = principal.getName() + ":" + extensionUniqueId;
+    public PrincipalIdentifierToken generateToken(Principal principal) throws NoSuchAlgorithmException, NoSuchPaddingException, InvalidKeyException, InvalidAlgorithmParameterException, IllegalBlockSizeException, BadPaddingException, IOException {
```

Comment:
Can you write tests for this to show its usage?

Reply:
Definitely!